### ISSUE TITLE
[MPS] Pooling backward SIGABRT crash on channels_last tensors (buffer too small)

### ISSUE THREAD

**Issue by npinto** [pytorch/pytorch#175190]
## Summary

`MaxPool2d`, `AvgPool2d`, and `AdaptiveAvgPool2d` backward on MPS crash with **SIGABRT** when the input is `channels_last`. The crash occurs in Metal's `MPSNDArray` — the buffer allocated for backward is sized for contiguous layout but channels_last requires more space.

```
MPSNDArray.mm:869: failed assertion
'[MPSNDArray, initWithBuffer:descriptor:] Error: buffer is not large enough. Must be 768 bytes'
```

Forward pass works correctly for all three ops. The crash is backward-only.

**AdaptiveMaxPool2d backward is NOT affected** (works correctly on channels_last).

Also crashes:
- `F.softmax` on certain non-contiguous shapes: MLIR reshape assertion (`'mps.reshape' op the result shape is not compatible with the input shape`)
- `BatchNorm3d` on non-contiguous 5D: `NSRangeException`

**Workaround:** `.contiguous()` before pooling.

## Reproducer (MaxPool2d backward crash)

**WARNING: This crashes the Python process with SIGABRT.**

```python
import torch
import torch.nn as nn
assert torch.backends.mps.is_available()

pool = nn.MaxPool2d(2)
x = torch.randn(4, 3, 8, 8).to("mps")
x = x.to(memory_format=torch.channels_last).requires_grad_(True)

out = pool(x)  # forward OK
out.sum().backward()  # SIGABRT: buffer not large enough
```

Same crash with `nn.AvgPool2d(2)` and `nn.AdaptiveAvgPool2d((4, 4))`.

## Reproducer (softmax crash)

**WARNING: This crashes the Python process.**

```python
import torch
import torch.nn.functional as F
assert torch.backends.mps.is_available()

x = torch.randn(1, 5, 5, 8, device="mps")
nc = x.permute(0, 3, 1, 2)  # (1, 8, 5, 5) non-contiguous
F.softmax(nc, dim=0)  # SIGABRT: MLIR reshape assertion
```

## Environment

- PyTorch 2.10.0
- macOS 15.4 (Sequoia)
- Apple M2 Max
- Python 3.13

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @xmfan @bobrenjc93 @mruberry @jbschlosser @walterddr @mikaylagawarecki @jamesr66a @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk
